homekit/StateManager.js

```diff
@@ -1,623 +1,1034 @@
-		smartModeState.highTemperatureState = {
-			on: false,
-			targetTemperature: device.state.targetTemperature,
-			temperatureUnit: device.temperatureUnit,
-			mode: device.state.mode,
-			fanSpeed: device.state.fanSpeed,
-			swing: device.state.verticalSwing,
-			horizontalSwing: device.state.horizontalSwing,
-			light: device.state.light
-		}
+		// If nothing has changed should we skip...? Like we do in StateHandler for SET?
 
-		smartModeState.lowTemperatureThreshold = device.state.targetTemperature - (device.usesFahrenheit ? 1.8 : 1)
-		smartModeState.lowTemperatureState = {
-			on: true,
-			targetTemperature: device.state.targetTemperature,
-			temperatureUnit: device.temperatureUnit,
-			mode: device.state.mode,
-			fanSpeed: device.state.fanSpeed,
-			swing: device.state.verticalSwing,
-			horizontalSwing: device.state.horizontalSwing,
-			light: device.state.light
+		if (!(device.state instanceof Classes.InternalAcState)) {
+			// TODO: log warning
+			return
 		}
-	}
 
-	// StateHandler is invoked as a Proxy, and therefore overwrites/intercepts the default get()/set() commands [traps]
-	// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
+		const smartModeState = device.state.smartMode
+
+		smartModeState.type = 'temperature'
+		smartModeState.highTemperatureWebhook = null
+		smartModeState.lowTemperatureWebhook = null
+
+		if (device.state.mode === 'COOL') {
+			smartModeState.highTemperatureThreshold = device.state.targetTemperature + (device.usesFahrenheit ? 1.8 : 1)*climateReactAutoSetupMultiplier + climateReactAutoSetupOffset
+			smartModeState.highTemperatureState = {
+				on: true,
+				targetTemperature: device.state.targetTemperature,
+				temperatureUnit: device.temperatureUnit,
+				mode: device.state.mode,
+				fanSpeed: device.state.fanSpeed,
+				swing: device.state.verticalSwing,
+				horizontalSwing: device.state.horizontalSwing,
+				light: device.state.light ? 'on' : 'off'
+			}
 
-	// NOTE: device.state is of "type" StateHandler. When one of its properties is "set" (e.g. device.state.<property> = <val>),
-	//       that's where we actually send commands to the appropriate Sensibo devices. If a property is not set, the aformentioned
-	//       code will not execute and the changes would not take effect.
-	//
-	//       For example, if we set a property of smartMode directly, e.g. device.state.smartMode.enabled = true, StateHandler's
-	//       setter will not get called and so any changes will not take effect. This is why we MUST update a device's property as
-	//       a whole, and do it only once (otherwise's the setter will get called multiple times which will send repeated commands
-	//       to the Sensibo devices).
-	device.state.smartMode = smartModeState
-}
+			smartModeState.lowTemperatureThreshold = device.state.targetTemperature - (device.usesFahrenheit ? 1.8 : 1)*climateReactAutoSetupMultiplier + climateReactAutoSetupOffset
```

Comment:
Should this be:
```suggestion
			smartModeState.lowTemperatureThreshold = device.state.targetTemperature - ((device.usesFahrenheit ? 1.8 : 1) * climateReactAutoSetupMultiplier + climateReactAutoSetupOffset)
```
Otherwise wouldn't the Offset _increase_ rather than decrease the threshold?

Reply:
Nope, this is fine. I'll explain the rational - the point of the offset is to allow something like this:
For example, when setting cool 25C, instead of keeping the temperature between 24 and 26, with an offset of 1, you can have it kept between 25 and 27. I actually use that for that exact purpose because I rather the AC stop cooling immediately when it reaches below 25 but also don't want it to start it at 26 because then it starts and stops every 5 minutes which drives me crazy. I rather have it start at 27 because then it only starts and stops every 10-15 minutes or so.

Reply:
I _think_ I now understand what you are aiming for, but it probably needs to be reworded to ensure others understand it (especially as I didn't the first time).

I'm probably struggling the most with 'multiplier'... maybe the settings should be called range and offset instead?
Offset would be defined as degrees above (or below) current AC temp to start (or stop), allowing your use case of 0 offset (and mine of -1)
Range would be defined as degrees +/- from that starting point to set the other end of the threshold.

I'm imagining:
Mode Cool, temp 25
Offset = 0
Range = 2
lowTemperatureThreshold = 25 + 0 (offset) (range not used) = 25
highTemperatureThreshold = 25 + 0 (offset) + 2 (range) = 27
AC turns on (set to 25) at 25 and turns off at 27

Mode Cool, temp 24
Offset = -1
Range = 2
lowTemperatureThreshold = 24 + -1 (range not used) = 23
highTemperatureThreshold = 24 + -1 + 2 = 25
AC turns on (set to 24) at 23 and turns off at 25

Mode Heat, temp 20
Offset = 0
Range = -3 (for Heat mode we could "simplify" it and have the negative implied, saves users needing to change their settings between seasons)
lowTemperatureThreshold = 20 + 0 + -3 = 17
highTemperatureThreshold = 20 + 0 (range not used) = 20
AC turns on (set to 20) at 17 and turns off at 20

Mode Heat, temp 22
Offset = 1
Range = -2 
lowTemperatureThreshold = 22 + 1 + -2 = 21
highTemperatureThreshold = 22 + 1 (range not used) = 23
AC turns on (set to 22) at 21 and turns off at 23

Reply:
Also still need to check how Sensibo API handles F degrees in "SmartMode"... i.e. do we need to change F to C